src/cryptography/hazmat/bindings/openssl/binding.py

```diff
@@ -139,3 +140,19 @@ def _lock_cb(cls, mode, n, file, line):
                     mode, n, file, line
                 )
             )
+
+
+# https://github.com/pyca/cryptography/issues/2299 contains extensive
+# discussion about this workaround. In essence, we are acquiring a global lock
+# (including across all subinterpreters if we're in a mod_wsgi environment)
+# initializing some C level locks, then initializing the binding and adding the
+# OS random engine. This should prevent engine addition race conditions,
+# although it's still unclear what happens when a subinterpreter which
+# has registered the locks is destroyed. Do the callback functions still work
+# or are all the objects associated with that subinterpreter destroyed?
+try:
```

Comment:
Ah, ok, so in 3.4/3.5, these refer to the real global lock.

The comment should mention this.

On Sun, Sep 13, 2015 at 4:02 PM, Donald Stufft <notifications@github.com>
wrote:

> In src/cryptography/hazmat/bindings/openssl/binding.py
> <https://github.com/pyca/cryptography/pull/2336#discussion_r39351421>:
>
> > @@ -139,3 +140,19 @@ def _lock_cb(cls, mode, n, file, line):
> >                      mode, n, file, line
> >                  )
> >              )
> > +
> > +
> > +# https://github.com/pyca/cryptography/issues/2299 contains extensive
> > +# discussion about this workaround. In essence, we are acquiring a global lock
> > +# (including across all subinterpreters if we're in a mod_wsgi environment)
> > +# initializing some C level locks, then initializing the binding and adding the
> > +# OS random engine. This should prevent engine addition race conditions,
> > +# although it's still unclear what happens when a subinterpreter which
> > +# has registered the locks is destroyed. Do the callback functions still work
> > +# or are all the objects associated with that subinterpreter destroyed?
> > +try:
>
> I guess that 3.4+'s imp.acquire_lock() is still global and isn't per
> module as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2336/files#r39351421>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084